### PR TITLE
Reduces the amount of hard drive memory required to store ensembles.

### PR DESCRIPTION
**Pull request by gui-miotto**
This is a follow up of PR #710 .

### CODE REVIEW

File: autosklearn/ensemble_builder.py [hunk 1]
```diff
@@ -195,6 +202,22 @@ def main(self):
                 continue
 
             selected_models = self.get_n_best_preds()
+            total_models = list(range(self.num_run))
+            to_be_removed_list = set(total_models) - set(self.best_model_index) - set(self.removed_model)
+
+            for file in to_be_removed_list:
+                # not sure about having zero as a constant
```
Comment (gui-miotto) [inline autosklearn/ensemble_builder.py]: @mfeurer , what do you say about this comment?

Comment (gui-miotto) [reply autosklearn/ensemble_builder.py]: Ignore the question for now. (I substituted 0 for the seed)

File: autosklearn/automl.py [hunk 2]
```diff
@@ -626,7 +628,7 @@ def fit_ensemble(self, y, task=None, metric=None, precision='32',
 
     def _get_ensemble_process(self, time_left_for_ensembles,
                               task=None, metric=None, precision=None,
-                              dataset_name=None, max_iterations=-1,
+                              dataset_name=None, max_iterations=None,
```
Comment (gui-miotto) [inline autosklearn/automl.py]: I changed this, because I didn't understand what would be the behavior for max_iterations=-1. As far as I understood, the value that is used to deactivate the max_iterations limit is None.

File: test/test_automl/test_estimators.py [hunk 3]
```diff
@@ -21,9 +21,10 @@
 from autosklearn.automl import AutoMLClassifier, AutoML
 from autosklearn.util.backend import Backend, BackendContext
 from autosklearn.constants import BINARY_CLASSIFICATION
-from base import Base
 
 sys.path.append(os.path.dirname(__file__))
+from base import Base
```
Comment (gui-miotto) [inline test/test_automl/test_estimators.py]: Shouldn't this be the correct order of these two lines of code?

Comment (mfeurer) [reply test/test_automl/test_estimators.py]: I guess so.

**Comment by gui-miotto** [comment]
@mfeurer I think you can start reviewing the PR.
The test with Python 3.7 is failing on the np.newaxis thing. But I don't think I have the permission to rerun it.
I still didn't rename selected_models to candidate_models. Gonna do that as the last step.

File: test/test_automl/test_automl.py [hunk 4]
```diff
@@ -114,6 +115,33 @@ def test_fit(self):
...
+            time_left_for_this_task=20,
+            per_run_time_limit=5,
+            ensemble_nbest=3,
+        )
+        automl.fit(
+            X_train, Y_train, metric=accuracy, task=MULTICLASS_CLASSIFICATION,
+        )
+
+        model_files = glob.glob(os.path.join(
+            backend_api.temporary_directory, '.auto-sklearn', 'models', '*.model'))
```
Comment (mfeurer) [inline test/test_automl/test_automl.py]: Isn't there a function in the backend to get the relevant model identifiers?

File: test/test_automl/test_automl.py [hunk 5]
```diff
@@ -114,6 +115,33 @@ def test_fit(self):
...
+
+        # Assert that the files of the models used by the ensemble are available
+        ensemble_members = set([idx[1] for idx in automl.ensemble_.identifiers_])
```
Comment (mfeurer) [inline test/test_automl/test_automl.py]: Could you please rename the variables so that they indicate that these are the indices? This would make the code a bit easier to understand.

File: test/test_automl/test_automl.py [hunk 6]
```diff
@@ -114,6 +115,33 @@ def test_fit(self):
...
+        model_files = set([int(m_file.split('.')[-2]) for m_file in model_files])
+        self.assertSetEqual(ensemble_members & model_files, ensemble_members)
```
Comment (mfeurer) [inline test/test_automl/test_automl.py]: Could you please evaluate whether it's possible to check that some files are deleted? Maybe with this wrap method? https://stackoverflow.com/a/25608441

File: autosklearn/automl.py (hunk 2 above)
Comment (mfeurer) [reply autosklearn/automl.py]: Yes, I agree.

File: test/test_automl/test_automl.py (hunk 5 above)
Comment (gui-miotto) [reply test/test_automl/test_automl.py]: Ok

File: test/test_automl/test_automl.py (hunk 4 above)
Comment (gui-miotto) [reply test/test_automl/test_automl.py]: There is indeed: list_all_models()
Let's use it then.

File: test/test_automl/test_automl.py (hunk 6 above)
Comment (gui-miotto) [reply test/test_automl/test_automl.py]: I tried to do this using a wrap, but was stumbling on some problems. The issue is that the EmsembleBuilder inside AutoML (automl.proc_ensemble) is purged at the end of the fit() function. I could get around this, but I would have to add some strange code just to make the test/wrapping work. Therefore, I opted by another more simple solution.

Since I'm logging the model deletions, I think the easiest way to check for deletions would be to inspect the log. That is what I implemented.

Let me know what your thoughts on this.
